// @angular-material-components/moment-adapter/bundles/angular-material-components-moment-adapter.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/core'), require('moment'), require('@angular-material-components/datetime-picker')) :
    typeof define === 'function' && define.amd ? define('@angular-material-components/moment-adapter', ['exports', '@angular/core', '@angular/material/core', 'moment', '@angular-material-components/datetime-picker'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['angular-material-components'] = global['angular-material-components'] || {}, global['angular-material-components']['moment-adapter'] = {}), global.ng.core, global.ng.material.core, global._rollupMoment, global['@angular-material-components/datetime-picker']));
}(this, (function (exports, core, core$1, _rollupMoment, datetimePicker) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var _rollupMoment__default = /*#__PURE__*/_interopDefaultLegacy(_rollupMoment);
    var _rollupMoment__namespace = /*#__PURE__*/_interopNamespace(_rollupMoment);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                __createBinding(exports, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /** @type {?} */
    var moment = _rollupMoment__default['default'] || _rollupMoment__namespace;
    /**
     * Configurable options for {\@see MomentDateAdapter}.
     * @record
     */
    function NgxMatMomentDateAdapterOptions() { }
    if (false) {
        /**
         * When enabled, the dates have to match the format exactly.
         * See https://momentjs.com/guides/#/parsing/strict-mode/.
         * @type {?|undefined}
         */
        NgxMatMomentDateAdapterOptions.prototype.strict;
        /**
         * Turns the use of utc dates on or off.
         * Changing this will change how Angular Material components like DatePicker output dates.
         * {\@default false}
         * @type {?|undefined}
         */
        NgxMatMomentDateAdapterOptions.prototype.useUtc;
    }
    /**
     * InjectionToken for moment date adapter to configure options.
     * @type {?}
     */
    var NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS = new core.InjectionToken('NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS', {
        providedIn: 'root',
        factory: NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS_FACTORY
    });
    /**
     * \@docs-private
     * @return {?}
     */
    function NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS_FACTORY() {
        return {
            useUtc: false
        };
    }
    /**
     * Creates an array and fills it with values.
     * @template T
     * @param {?} length
     * @param {?} valueFunction
     * @return {?}
     */
    function range(length, valueFunction) {
        /** @type {?} */
        var valuesArray = Array(length);
        for (var i = 0; i < length; i++) {
            valuesArray[i] = valueFunction(i);
        }
        return valuesArray;
    }
    /**
     * Adapts Moment.js Dates for use with Angular Material.
     */
    var NgxMatMomentAdapter = /** @class */ (function (_super) {
        __extends(NgxMatMomentAdapter, _super);
        /**
         * @param {?} dateLocale
         * @param {?=} _options
         */
        function NgxMatMomentAdapter(dateLocale, _options) {
            var _this = _super.call(this) || this;
            _this._options = _options;
            _this.setLocale(dateLocale || moment.locale());
            return _this;
        }
        /**
         * @param {?} locale
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.setLocale = function (locale) {
            var _this = this;
            _super.prototype.setLocale.call(this, locale);
            /** @type {?} */
            var momentLocaleData = moment.localeData(locale);
            this._localeData = {
                firstDayOfWeek: momentLocaleData.firstDayOfWeek(),
                longMonths: momentLocaleData.months(),
                shortMonths: momentLocaleData.monthsShort(),
                dates: range(31, ( /**
                 * @param {?} i
                 * @return {?}
                 */function (i) { return _this.createDate(2017, 0, i + 1).format('D'); })),
                longDaysOfWeek: momentLocaleData.weekdays(),
                shortDaysOfWeek: momentLocaleData.weekdaysShort(),
                narrowDaysOfWeek: momentLocaleData.weekdaysMin(),
            };
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getYear = function (date) {
            return this.clone(date).year();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getMonth = function (date) {
            return this.clone(date).month();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getDate = function (date) {
            return this.clone(date).date();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getDayOfWeek = function (date) {
            return this.clone(date).day();
        };
        /**
         * @param {?} style
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getMonthNames = function (style) {
            // Moment.js doesn't support narrow month names, so we just use short if narrow is requested.
            return style == 'long' ? this._localeData.longMonths : this._localeData.shortMonths;
        };
        /**
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getDateNames = function () {
            return this._localeData.dates;
        };
        /**
         * @param {?} style
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getDayOfWeekNames = function (style) {
            if (style == 'long') {
                return this._localeData.longDaysOfWeek;
            }
            if (style == 'short') {
                return this._localeData.shortDaysOfWeek;
            }
            return this._localeData.narrowDaysOfWeek;
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getYearName = function (date) {
            return this.clone(date).format('YYYY');
        };
        /**
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getFirstDayOfWeek = function () {
            return this._localeData.firstDayOfWeek;
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getNumDaysInMonth = function (date) {
            return this.clone(date).daysInMonth();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.clone = function (date) {
            return date.clone().locale(this.locale);
        };
        /**
         * @param {?} year
         * @param {?} month
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.createDate = function (year, month, date) {
            // Moment.js will create an invalid date if any of the components are out of bounds, but we
            // explicitly check each case so we can throw more descriptive errors.
            if (month < 0 || month > 11) {
                throw Error("Invalid month index \"" + month + "\". Month index has to be between 0 and 11.");
            }
            if (date < 1) {
                throw Error("Invalid date \"" + date + "\". Date has to be greater than 0.");
            }
            /** @type {?} */
            var result = this._createMoment({ year: year, month: month, date: date }).locale(this.locale);
            // If the result isn't valid, the date must have been out of bounds for this month.
            if (!result.isValid()) {
                throw Error("Invalid date \"" + date + "\" for month with index \"" + month + "\".");
            }
            return result;
        };
        /**
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.today = function () {
            return this._createMoment().locale(this.locale);
        };
        /**
         * @param {?} value
         * @param {?} parseFormat
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.parse = function (value, parseFormat) {
            if (value && typeof value == 'string') {
                return this._createMoment(value, parseFormat, this.locale);
            }
            return value ? this._createMoment(value).locale(this.locale) : null;
        };
        /**
         * @param {?} date
         * @param {?} displayFormat
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.format = function (date, displayFormat) {
            date = this.clone(date);
            if (!this.isValid(date)) {
                throw Error('MomentDateAdapter: Cannot format invalid date.');
            }
            return date.format(displayFormat);
        };
        /**
         * @param {?} date
         * @param {?} years
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.addCalendarYears = function (date, years) {
            return this.clone(date).add({ years: years });
        };
        /**
         * @param {?} date
         * @param {?} months
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.addCalendarMonths = function (date, months) {
            return this.clone(date).add({ months: months });
        };
        /**
         * @param {?} date
         * @param {?} days
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.addCalendarDays = function (date, days) {
            return this.clone(date).add({ days: days });
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.toIso8601 = function (date) {
            return this.clone(date).format();
        };
        /**
         * Returns the given value if given a valid Moment or null. Deserializes valid ISO 8601 strings
         * (https://www.ietf.org/rfc/rfc3339.txt) and valid Date objects into valid Moments and empty
         * string into null. Returns an invalid date for all other values.
         * @param {?} value
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.deserialize = function (value) {
            /** @type {?} */
            var date;
            if (value instanceof Date) {
                date = this._createMoment(value).locale(this.locale);
            }
            else if (this.isDateInstance(value)) {
                // Note: assumes that cloning also sets the correct locale.
                return this.clone(value);
            }
            if (typeof value === 'string') {
                if (!value) {
                    return null;
                }
                date = this._createMoment(value, moment.ISO_8601).locale(this.locale);
            }
            if (date && this.isValid(date)) {
                return this._createMoment(date).locale(this.locale);
            }
            return _super.prototype.deserialize.call(this, value);
        };
        /**
         * @param {?} obj
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.isDateInstance = function (obj) {
            return moment.isMoment(obj);
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.isValid = function (date) {
            return this.clone(date).isValid();
        };
        /**
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.invalid = function () {
            return moment.invalid();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getHour = function (date) {
            return date.hours();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getMinute = function (date) {
            return date.minutes();
        };
        /**
         * @param {?} date
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.getSecond = function (date) {
            return date.seconds();
        };
        /**
         * @param {?} date
         * @param {?} value
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.setHour = function (date, value) {
            date.hours(value);
        };
        /**
         * @param {?} date
         * @param {?} value
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.setMinute = function (date, value) {
            date.minutes(value);
        };
        /**
         * @param {?} date
         * @param {?} value
         * @return {?}
         */
        NgxMatMomentAdapter.prototype.setSecond = function (date, value) {
            date.seconds(value);
        };
        /**
         * Creates a Moment instance while respecting the current UTC settings.
         * @private
         * @param {?} date
         * @param {?=} format
         * @param {?=} locale
         * @return {?}
         */
        NgxMatMomentAdapter.prototype._createMoment = function (date, format, locale) {
            var _a = this._options || {}, strict = _a.strict, useUtc = _a.useUtc;
            return useUtc
                ? moment.utc(date, format, locale, strict)
                : moment(date, format, locale, strict);
        };
        return NgxMatMomentAdapter;
    }(datetimePicker.NgxMatDateAdapter));
    NgxMatMomentAdapter.decorators = [
        { type: core.Injectable }
    ];
    /** @nocollapse */
    NgxMatMomentAdapter.ctorParameters = function () { return [
        { type: String, decorators: [{ type: core.Optional }, { type: core.Inject, args: [core$1.MAT_DATE_LOCALE,] }] },
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS,] }] }
    ]; };
    if (false) {
        /**
         * @type {?}
         * @private
         */
        NgxMatMomentAdapter.prototype._localeData;
        /**
         * @type {?}
         * @private
         */
        NgxMatMomentAdapter.prototype._options;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/moment-formats.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /** @type {?} */
    var DEFAULT_DATE_INPUT = 'l, LTS';
    /** @type {?} */
    var NGX_MAT_MOMENT_FORMATS = {
        parse: {
            dateInput: DEFAULT_DATE_INPUT,
        },
        display: {
            dateInput: DEFAULT_DATE_INPUT,
            monthYearLabel: 'MMM YYYY',
            dateA11yLabel: 'LL',
            monthYearA11yLabel: 'MMMM YYYY',
        },
    };

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/moment-adapter.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxMomentDateModule = /** @class */ (function () {
        function NgxMomentDateModule() {
        }
        return NgxMomentDateModule;
    }());
    NgxMomentDateModule.decorators = [
        { type: core.NgModule, args: [{
                    providers: [
                        {
                            provide: datetimePicker.NgxMatDateAdapter,
                            useClass: NgxMatMomentAdapter,
                            deps: [core$1.MAT_DATE_LOCALE, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS]
                        }
                    ],
                },] }
    ];
    var ɵ0 = NGX_MAT_MOMENT_FORMATS;
    var NgxMatMomentModule = /** @class */ (function () {
        function NgxMatMomentModule() {
        }
        return NgxMatMomentModule;
    }());
    NgxMatMomentModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [NgxMomentDateModule],
                    providers: [{ provide: datetimePicker.NGX_MAT_DATE_FORMATS, useValue: ɵ0 }],
                },] }
    ];

    /**
     * @fileoverview added by tsickle
     * Generated from: public-api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: angular-material-components-moment-adapter.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS = NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS;
    exports.NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS_FACTORY = NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS_FACTORY;
    exports.NGX_MAT_MOMENT_FORMATS = NGX_MAT_MOMENT_FORMATS;
    exports.NgxMatMomentAdapter = NgxMatMomentAdapter;
    exports.NgxMatMomentModule = NgxMatMomentModule;
    exports.NgxMomentDateModule = NgxMomentDateModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
