// ngx-mat-moment-adapter/esm2015/lib/ngx-mat-moment-adapter.module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { MAT_DATE_FORMATS, MAT_DATE_LOCALE } from '@angular/material/core';
import { MAT_MOMENT_DATE_ADAPTER_OPTIONS, NgxMatMomentAdapter } from './ngx-mat-moment-adapter';
import { NGX_MAT_MOMENT_FORMATS } from './ngx-mat-moment-formats';
import { NgxMatDateAdapter } from 'ngx-mat-datetime-picker';
let NgxMomentDateModule = class NgxMomentDateModule {
};
NgxMomentDateModule = __decorate([
    NgModule({
        providers: [
            {
                provide: NgxMatDateAdapter,
                useClass: NgxMatMomentAdapter,
                deps: [MAT_DATE_LOCALE, MAT_MOMENT_DATE_ADAPTER_OPTIONS]
            }
        ],
    })
], NgxMomentDateModule);
export { NgxMomentDateModule };
const ɵ0 = NGX_MAT_MOMENT_FORMATS;
let NgxMatMomentModule = class NgxMatMomentModule {
};
NgxMatMomentModule = __decorate([
    NgModule({
        imports: [NgxMomentDateModule],
        providers: [{ provide: MAT_DATE_FORMATS, useValue: ɵ0 }],
    })
], NgxMatMomentModule);
export { NgxMatMomentModule };
export { ɵ0 };
